mlflow/models/utils.py

```diff
@@ -906,15 +906,21 @@ def _enforce_datatype(data: Any, dtype: DataType):
 def _enforce_array(data: Any, arr: Array, required=True):
     if not required and data is None:
         return None
+
+    if len(data) == 0:
+        raise MlflowException(
+            "Expected a non-empty list/array for Array column. If you want an empty value, "
+            "mark the column `required=False` and put `None` instead."
+        )
```

Comment:
Here added check for empty list/array. `len()` doesn't catch nested list like `[[], []]` but it will be caught in recursive call anyway.

Reply:
What would happen if we accept an empty list/array?

Reply:
Not an issue of enforcing given schema on data indeed, while it causes inconsistency with `_infer_colspec_type`, which doesn't allow empty array (for inner-element of nested array as well).

However, on second thought, we should probably be able to allow empty array here. As far as I see [_infer_colspec_type code](https://github.com/mlflow/mlflow/blob/llm_signature/mlflow/types/utils.py#L121-L124), it should only block empty array at top level (sample level), rather than element in each sub-list, e.g. `{"data": []}` is not allowed but `{"data: [["a", "b"],  ["c", "d"], []]}` can be allowed, because we can infer the column type from the other elements. If we make this change, removing chech in `_enforce_array` won't cause inconsistency.

WDYT @harupy @serena-ruan ?

Reply:
Thanks for the explanation! 

1. I agree. They don't have to be the same. `_enforce_array` should accept an empty array. The model should decide what to do with it.
2. I agree. `_infer_colspec_type code` should work for `{"data": [["a", "b], []}` (unless it's super complex to implement it).

Reply:
Awesome! Will include fix for (1) in this PR and cut a separate one for (2)

Reply:
I don't think `_infer_colspec_type` should work for `{"data": [["a", "b], []]}`. It should be inferred as `Array(Array(string))`, and we have the assumption that elements in an array should be of same type. I think it could be None like `{"data": [["a", "b"], None, np.nan]}` for `ColSpec(Array(Array(string)), required=False)`, but empty list is not omitting the value.

Reply:
For 1: from my opinion I think it shouldn't pass for `_enforce_array` on an empty array, but let's check what's the behavior for 2.7.1 and keep it.

Reply:
IMO empty list is different from None, the former represents no element while the latter is null.

For example, let's say there is a column listing name of students in a club
```
{"list_of_students": []}
```
says the club doesn't have any member. However,
```
{"list_of_students": None}
```
says no info on the column -> optional.

Reply:
@harupy @serena-ruan 
Thank you for providing your thoughts! It seems both perspective have some legit points, but I have to get a consensus to close this PR;

For 1, apparently [_enforce_mlflow_datatype](https://github.com/mlflow/mlflow/blame/v2.7.1/mlflow/models/utils.py#L382-L388) in 2.7.1 does nothing for an empty list or a nested list includes an empty list, as it basically avoid touching object type much. I'm not 100% sure and please correct me if I'm wrong, but if this is the case, can we align on allowing empty list here?

For 2, I think the key point is whether we are ok with this assumption

> we have the assumption that elements in an array should be of same type.

I think assuming type based on non-empty values is fine, and we can still throw exception when the inputs are all empty list. The opposite (throw for any empty list) sounds a bit too strict, particularly on LLM use case, as it's easy to mix some empty list in huge semi-structured dataset like scraped documents.

Reply:
> infer_signature([]) gives ColSpec(string)

This does seem strange. Could be an uncaught bug.

Reply:
I think it's because we use `all()` here, which returns True for empty list.
https://github.com/mlflow/mlflow/blob/v2.7.1/mlflow/types/utils.py#L150

Reply:
@B-Step62 Thanks for the investigation. I think `infer_signature([])` should throw.

Reply:
@harupy Agreed, should we fix on master branch? It behaves fine in this branch (throws for []) so probably no need for now as long as we merge this soon.

Reply:
If it behaves fine, we don't need to.